spring-cloud-gcp-storage/src/main/java/com/google/cloud/spring/storage/GoogleStorageResource.java

```diff
@@ -134,27 +134,38 @@ public boolean isOpen() {
 	}
 
 	/**
-	 * Since the gs: protocol will normally not have a URL stream handler registered,
-	 * this method will always throw a {@link java.net.MalformedURLException}.
-	 * @return the URL for the GCS resource, if a URL stream handler is registered for the gs protocol.
+	 * @return the self-link for this GCS resource. Note that this is not a URL to download the contents of the file.
 	 */
 	@Override
+	@NonNull
 	public URL getURL() throws IOException {
-		return getURI().toURL();
+		String url;
+		if (isBucket()) {
+			Bucket bucket = getBucket();
+			Assert.notNull(bucket, "The bucket " + this.getBucketName() + " does not exist.");
+			url = bucket.getSelfLink();
+		}
+		else {
+			Blob b = getBlob();
+			Assert.notNull(b, "The object at " + this.getURI() + " does not exist.");
+			url = getBlob().getSelfLink();
+		}
+		return new URL(url);
 	}
 
 	@Override
+	@NonNull
 	public URI getURI() {
 		return this.location.uri();
```

Comment:
Hmmm, URI will be different than URL, which could be confusing. Maybe we should document this a bit.

Reply:
From https://www.guru99.com/url-vs-uri-difference.html#:~:text=URI-,URL%20stands%20for%20Uniform%20Resource%20Locator.,Uniform%20Resource%20Name)%20or%20both.:

> Not all URIs are URLs since a URI can be a name instead of a locator.

In this case, it's URI doesn't conform to a URL syntax, so the self-link is the best we can do without adding some hacky reflection-based `UrlStreamHandlerFactory` impl.

Reply:
Yeah I agree with the workaround. I just think it would be helpful to document it.